File: 0901-1000/0950-Reveal Cards In Increasing Order/0950-Reveal Cards In Increasing Order.py
import collections
# Do reverse operation
class Solution:
    def deckRevealedIncreasing(self, deck: List[int]) -> List[int]:
        deck.sort()
        for i in range(len(deck) - 2, -1, -1):
            deck[i + 1:] = deck[-1:] + deck[i + 1: -1]
        return deck

# use queue to optimize time complexity              
class Solution2:
    def deckRevealedIncreasing(self, deck: List[int]) -> List[int]:
        deck.sort()
        Q = collections.deque([deck[-1]])
        for i in range(len(deck) - 2, -1, -1):
            Q.appendleft(Q.pop())
            Q.appendleft(deck[i])
        return list(Q)

class Solution3:
    def deckRevealedIncreasing(self, deck: List[int]) -> List[int]:
        Q = collections.deque(range(len(deck)))
        result = [None] * len(deck)
        for card in sorted(deck):
            result[Q.popleft()] = card
            if Q:
                Q.append(Q.popleft())
        return result

File: 0901-1000/0950-Reveal Cards In Increasing Order/0950-Reveal Cards In Increasing Order.cpp
class Solution {
public:
    vector<int> deckRevealedIncreasing(vector<int>& deck) {
        sort(deck.begin(), deck.end());
        deque<int> Q;
        Q.push_back(deck.back());
        for (int i = (int)(deck.size()) - 2; i >= 0; --i) {
            Q.push_front(Q.back());
            Q.pop_back();
            Q.push_front(deck[i]);
        }
        return vector<int>(Q.begin(), Q.end());
    }
};

class Solution {
public:
    vector<int> deckRevealedIncreasing(vector<int>& deck) {
        sort(deck.begin(), deck.end());
        queue<int> Q;
        for (int i = 0; i < deck.size(); ++i) {
            Q.push(i);
        }
        vector<int> result(deck.size());
        for (int card : deck) {
            result[Q.front()] = card;
            Q.pop();
            if (Q.size()) {
                Q.push(Q.front());
                Q.pop();
            }
        }
        return result;
    }
};
